client/src/components/order/new-order.js

```diff
@@ -0,0 +1,279 @@
+    Api.get(`tables/available?id=${match.params.id}&start=${start}&finish=${finish}`)
+      .then((response) => {
+        if (response.error) {
+          // eslint-disable-next-line no-console
+          console.error(response);
+          return;
+        }
+        this.setState({
+          availableTables: response.data,
+        });
+      });
+  }
+
+  getTableCapacity(id) {
+    const table = this.state.availableTables.find((t) => {
+      return t.id === +id;
+    });
+    if (table) {
+      return table.capacity;
+    }
+    return null;
+  }
+
+  getMaximumOfVisitors() {
+    if (this.state.table) {
+      return this.getTableCapacity(this.state.table);
+    }
+    return this.state.availableTables.length && this.state.availableTables[0].capacity;
+  }
+
+  saveOrder() {
+    const { match, history } = this.props;
+    let tableId;
+    if (this.state.table) {
+      tableId = this.state.table;
+    } else if (this.state.availableTables.length) {
+      tableId = this.state.availableTables[0].id;
+    } else {
+      return;
+    }
+    const order = {
+      person: { id: 1 },
+      startTime: this.state.startDate.toUTCString(),
+      finishTime: this.state.finishDate.toUTCString(),
+      visitors: this.state.visitors,
+      table: { id: tableId },
+      description: this.state.description,
+    };
+
+    Api.post('orders', order)
+      .then((response) => {
+        if (response.error) {
+          // eslint-disable-next-line no-console
+          console.error(response);
+          this.setState({
+            isBadRequestError: true,
+          });
+          return;
+        }
+        history.push(match.url.replace(this.path, ''));
+      });
+  }
+
+  selectDateTime(label) {
+    return (
+      <DatePicker
+        selected={this.state[label]}
+        onChange={(date) => this.handleDateTimePicker(label, date)}
+        showTimeSelect
+        timeFormat="HH:mm"
+        className="date-time-input"
+        minDate={this.currentDate}
+        timeIntervals={this.timeInterval}
+        timeCaption="time"
+        dateFormat="yyyy-MM-dd HH:mm"
+      />
+    );
+  }
+
+  timeFormatter(time) {
+    if (time.getMinutes() % this.timeInterval) {
+      const intervalCount = time.getMinutes() / this.timeInterval;
+      const formattedMinutes = (intervalCount - (intervalCount % 1) + 1) * this.timeInterval;
+      time.setMinutes(formattedMinutes);
+    }
+    return time;
+  }
+
+  handleDateTimePicker(label, date) {
+    this.setState({
+      [label]: date,
+    }, () => {
+      this.getAvailableTables();
+    });
+  }
+
+  handleFormControl(data) {
+    this.setState({
+      [data.target.name]: data.target.value,
+    });
+  }
+
+  formatDateTime(date) {
+    return `${date.getFullYear()}-${date.getMonth() + 1}-${date.getDate()} ${date.getHours()}:${date.getMinutes()}`;
+  }
+
+  isDateTimesInvalid() {
+    return this.state.startDate >= this.state.finishDate;
+  }
```

Comment:
I think it is better to do this check when 'Create new order' button is clicked. I tried this and find UX can be better. 

Consider you pick start date and time for next week's Saturday on 8 p.m. and available finish date will be also next Saturday and starting 8:15 pm. This will let to alert '_Finish time should be **after** start time!_' only when user wants to pick finish date that is 'before' start date.

Reply:
Button `Save order` is disable when form is invalid. User can't click it when form is invalid

Reply:
I was mistaked. I had to write this:

> I think it is better to do this check when **inactive finish date on datepicker** is choosen before start date. I tried this and find UX can be better.

Reply:
> finish date on datepicker is choosen before start date

I don't understand. And in any cases, `finish date` must be chose only after `start date`. People think from start to finish not vice versa.

Reply:
I'm not what are you talking about. I'm not saying that _People think from start to finish not vice versa._ I mean alerts, we should change their behavior. 

> Consider you pick start date and time for next week's Saturday on 8 p.m. 

**now alert "Finish time should be more than Start time" is shown**. But: user haven't choosen finish date yet. 

Imagine that we have no any alert in case above.

**Case 1:** Then, user chooses available finish date - also next Saturday and starting 8:15 pm. It is ok due to time rules and user won't see any alerts anymore.

**Case 2:** But, when he wants to choose finish date that in fact is before his choosen start date, here we should show **alert**. Moreover, for better UX we should make dates 'before' choosen start inactive, but still show alert if user tries to pick them.

Thus, it will let to alert "Finish time should be after start time!" only in one case - when user wants to pick finish date that is 'before' start date.

Reply:
Current alert teach user how to use our application.

> Then, user chooses available finish date

What happen if user doesn't choose finish date? He or she doesn't see any alert and doesn't click Save button (form is invalid)

> Moreover, for better UX we should make dates 'before' choosen start inactive

I couldn't do it in this datapicker https://github.com/Hacker0x01/react-datepicker/issues/2077 and other

Reply:
It is clear that alert should be shown if smth goes wrong. When user picks start date, he hasn't done smth wrong to see alert. We can move finish date due to choosen start date and + 15 minutes, for example. Now user can understand that he isn't able to use finish time 'before' start. 

Also dates before finish time should be inactive and we shall do it even in another, special task. Otherwise we allow user to do abnormal thing in our app.

Also we shold restrict user to make order starting 20.04.20 and ending 01.05.20 as restaurant will be closed from 24:00 to 12:00, for example.

This should be marked to fix in another tasks or may be this.

Reply:
> We can move finish date due to choosen start date and + 15 minutes, for example. 

I've tried to do it, but couldn't

> Also dates before finish time should be inactive

...
> I couldn't do it in this datapicker Hacker0x01/react-datepicker#2077 and other

> Also we shold restrict user to make order starting 20.04.20 and ending 01.05.20 as restaurant will be closed from 24:00 to 12:00, for example.

It's a good idea, I think, you should create new task

Reply:
> Also dates before finish time should be inactive

Sorry, it should be this:
Also dates and times before start time should be inactive.

> It's a good idea, I think, you should create new task

We can create this task only whe this branch will be merged.

Reply:
> Sorry, it should be this:
Also dates and times before start time should be inactive.

...
> I couldn't do it in this datapicker Hacker0x01/react-datepicker#2077 and other

Reply:
According to this inactive 'days' presense, I guess it is possible to implement inactive status. Let us do it in another task.

![image](https://user-images.githubusercontent.com/50665431/79740380-3fb49e00-8308-11ea-9605-e40594546976.png)

Reply:
Did you read issue? Making date is inactive is possible, but not time

Reply:
In your code not only time but date can be choosen earlier than start date.
We can fix time but still can choose finish date before start, so we need to fix both.

Reply:
Server handles this case and client shows message if request is wrong. It isn't problem